src/main/java/org/prebid/server/cache/CacheService.java

```diff
@@ -326,16 +330,47 @@ private static PutObject createJsonPutObjectOpenrtb(CacheBid cacheBid) {
     /**
      * Makes XML type {@link PutObject} from {@link com.iab.openrtb.response.Bid}. Used for OpenRTB auction request.
      */
-    private static PutObject createXmlPutObjectOpenrtb(CacheBid cacheBid) {
-        if (cacheBid.getBid().getAdm() == null) {
-            return PutObject.of("xml", new TextNode("<VAST version=\"3.0\"><Ad><Wrapper>"
+    private static PutObject createXmlPutObjectOpenrtb(CacheBid cacheBid, List<String> bidIdsToModify,
+                                                       String accountId) {
+        final com.iab.openrtb.response.Bid bid = cacheBid.getBid();
+        String stringValue;
+        if (bid.getAdm() == null) {
+            stringValue = "<VAST version=\"3.0\"><Ad><Wrapper>"
                     + "<AdSystem>prebid.org wrapper</AdSystem>"
-                    + "<VASTAdTagURI><![CDATA[" + cacheBid.getBid().getNurl() + "]]></VASTAdTagURI>"
+                    + "<VASTAdTagURI><![CDATA[" + bid.getNurl() + "]]></VASTAdTagURI>"
                     + "<Impression></Impression><Creatives></Creatives>"
-                    + "</Wrapper></Ad></VAST>"), cacheBid.getTtl());
+                    + "</Wrapper></Ad></VAST>";
         } else {
-            return PutObject.of("xml", new TextNode(cacheBid.getBid().getAdm()), cacheBid.getTtl());
+            stringValue = bid.getAdm();
         }
+
+        final String bidId = bid.getId();
+        if (CollectionUtils.isNotEmpty(bidIdsToModify) && bidIdsToModify.contains(bidId)) {
+            stringValue = modifyVastXml(stringValue, bidId, accountId);
+        }
+
+        return PutObject.of("xml", new TextNode(stringValue), cacheBid.getTtl());
+    }
+
+    private static String modifyVastXml(String stringValue, String bidId, String accountId) {
+        final String closeTag = "</Impression>";
+        final int closeTagIndex = stringValue.indexOf(closeTag);
+
+        // no impression tag - pass it as it is
+        if (closeTagIndex == -1) {
+            return stringValue;
+        }
+
+        final String impressionUrl = String.format("https://prebid-server.rubiconproject.com/event?t=imp&b=%s&f=b&a=%s",
+                bidId, accountId);
+        final String openTag = "<Impression>";
+
+        // empty impression tag - just insert the link
+        if (closeTagIndex - stringValue.indexOf(openTag) == openTag.length()) {
+            return stringValue.replaceFirst(openTag, openTag + impressionUrl);
+        }
+
+        return stringValue.replaceFirst(closeTag, closeTag + "\n" + openTag + impressionUrl + closeTag);
```

Comment:
Do we really need `\n` for XML?

Reply:
I thought it would be good for a better styling, or at least that's how Bob recommended doing it.

Doesn't harm and makes up for a clearer xml, but I can remove it (an it will make this line clearer) if you think it's unnecessary. As you say.

Reply:
i think we don't need it.